inference-gateway/src/events_reporter.rs

```diff
@@ -0,0 +1,146 @@
+use anyhow::{Context, Result};
+use chrono::{DateTime, Utc};
+use log::info;
+use pgmq::PGMQueueExt;
+use serde::{Deserialize, Serialize};
+use sqlx::{Pool, Postgres};
+use std::{env, time::Duration};
+use tokio::time::interval;
+use uuid::Uuid;
+
+use crate::db::connect;
+use crate::errors::DatabaseError;
+
+pub fn split_events(events: Vec<Events>, max_size: usize) -> Vec<Message> {
+    events
+        .chunks(max_size)
+        .map(|chunk| Message {
+            id: Uuid::new_v4().to_string(),
+            message: chunk.to_vec(),
+        })
+        .collect()
+}
+
+pub async fn get_usage(
+    dbclient: &Pool<Postgres>,
+    start_time: DateTime<Utc>,
+    end_time: DateTime<Utc>,
+) -> Result<Vec<Events>, DatabaseError> {
+    let rows = sqlx::query_as!(
+        UsageData,
+        r#"
+        SELECT
+          organization_id,
+          instance_id,
+          duration_ms,
+          prompt_tokens,
+          completion_tokens,
+          model,
+          completed_at
+        FROM inference.requests
+        WHERE
+          completed_at >= $1 AND completed_at < $2
+        "#,
+        start_time,
+        end_time
+    )
+    .fetch_all(dbclient)
+    .await?;
+
+    Ok(rows_to_events(rows))
+}
+
+pub fn rows_to_events(rows: Vec<UsageData>) -> Vec<Events> {
+    rows.into_iter()
+        .map(|row| Events {
+            idempotency_key: Uuid::new_v4().to_string(),
+            organization_id: row.organization_id,
+            instance_id: row.instance_id,
+            payload: Payload {
+                completed_at: row.completed_at.to_string(),
+                duration_ms: row.duration_ms,
+                model: row.model,
+                prompt_tokens: row.prompt_tokens.to_string(),
+                completion_tokens: row.completion_tokens.to_string(),
+            },
+        })
+        .collect()
+}
+
+pub async fn events_reporter() -> Result<()> {
+    const BATCH_SIZE: usize = 1000;
+
+    let pg_conn_url = env::var("POSTGRES_QUEUE_CONNECTION")
+        .with_context(|| "POSTGRES_QUEUE_CONNECTION must be set")?;
+    let dbclient = connect(&pg_conn_url, 5).await?;
+
+    let queue = PGMQueueExt::new(pg_conn_url, 5).await?;
+
+    // TODO: Need to set this env variable
+    let metrics_events_queue =
+        env::var("BILLING_EVENTS_QUEUE").expect("BILLING_EVENTS_QUEUE must be set");
+
+    queue.init().await?;
+    queue.create(&metrics_events_queue).await?;
+
+    let mut sync_interval = interval(Duration::from_secs(3600));
+
+    loop {
+        sync_interval.tick().await;
+
+        let end_time = Utc::now();
+        let start_time = end_time - chrono::Duration::hours(1);
```

Comment:
Instead of doing now, minus and hour, I think we can do "for each of the previous hours in the last 24 hours not yet reported... then report the billing for that hour".

To remember which hours have already been sent, it can remember the idempotency keys in the AI database new table. The idempotency key can represent an hour of billing data. Also it should not send any hour of billing data for which the hour is not yet over.